### PR TITLE
Don't load initial memory state via code when using continuations.

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

**Review by georgwiese — DISMISSED** [review]
Cool! LGTM, just some minor comments.

File: riscv/src/continuations.rs [hunk 1]
```diff
@@ -152,10 +186,24 @@ pub fn rust_continuations_dry_run<F: FieldElement>(
...
+    // Get initial memory contents from the special variable "initial_memory".
+    // In the first full run, we use it as the memory contents of the executor;
+    // on the independent chunk runs, the executor uses zeroed initial memory,
+    // and the pages are loaded via the bootloader.
+    let initial_memory = load_initial_memory(&program);
+
+    let mut merkle_tree = MerkleTree::<F>::new();
+    merkle_tree.update(initial_memory.iter().map(|(k, v)| (*k, *v)));
+
+    // TODO: commit to the merkle_tree root in the verifier.
```
[the remark this answers is not in the document]
Comment (lvella) [reply riscv/src/continuations.rs:200]: What you didn't understand in the comment?

What about an `MerkleTree::from_initial_memory(ast)`?

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -418,6 +450,10 @@ machine Main {{
 
 {}
 
+let initial_memory = [
+{}
+];
```
[the remark this answers is not in the document]
Comment (lvella) [reply riscv/src/compiler.rs:455]: Done.

File: riscv/src/continuations.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lvella) [reply riscv/src/continuations.rs:200]: Let's leave this to a follow up PR.

File: riscv/src/continuations.rs [hunk 3]
```diff
@@ -152,10 +186,24 @@ pub fn rust_continuations_dry_run<F: FieldElement>(
     let program = pipeline.compute_analyzed_asm().unwrap().clone();
     sanity_check(&program);
 
+    log::info!("Initializing memory merkle tree...");
+
+    // Get initial memory contents from the special variable "initial_memory".
+    // In the first full run, we use it as the memory contents of the executor;
+    // on the independent chunk runs, the executor uses zeroed initial memory,
+    // and the pages are loaded via the bootloader.
```
Comment (leonardoalt) [inline riscv/src/continuations.rs:194]: So the main difference is that now the bootloader takes the [potentially] non-zero initial memory into account when computing the values of the used pages for the first chunks? The bootloader still loads the used memory pages for the first chunk, right?

File: riscv/src/compiler.rs [hunk 4]
```diff
@@ -144,34 +144,59 @@ pub fn compile(
...
-                    "mstore wrap(tmp1 - tmp2);".to_string(),
-                ]);
-                */
+            match value {
+                SingleDataValue::Value(v) => {
+                    if with_bootloader {
+                        // Instead of generating the data loading code, we store it
+                        // in the variable that will be used as the initial memory
+                        // snapshot, committed by the bootloader.
+                        initial_mem.push(format!("(0x{addr:x}, 0x{v:x})"));
```
Comment (leonardoalt) [inline riscv/src/compiler.rs:166]: why is this pushed as a string if the values are already available?

File: riscv/src/continuations.rs (hunk 3 above)
Comment (lvella) [reply riscv/src/continuations.rs:194]: > The bootloader still loads the used memory pages for the first chunk, right?
The bootloader wasn't loading anything in the first chunk. Now it is.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**